language-support/java/codegen/src/it/java/com/daml/ContractKeysTest.java

```diff
@@ -103,4 +104,21 @@ void nestedTupleKeyIsPartyAndInt() {
     assertEquals(nestedTupleKey.key.get()._2._3, "foobar");
     assertEquals(nestedTupleKey.key.get()._2._4.longValue(), 0L);
   }
+
+  @Test
+  void keyJsonDecoder() {
+    JsonLfDecoder<String> pk = PartyKey.Contract.keyJsonDecoder();
+    JsonLfDecoder<PartyAndInt> rk = RecordKey.Contract.keyJsonDecoder();
+    JsonLfDecoder<Tuple2<String, Long>> tk = TupleKey.Contract.keyJsonDecoder();
+    JsonLfDecoder<Tuple2<Tuple3<String, Long, String>, Tuple4<Long, Boolean, String, Long>>> ntk =
+        NestedTupleKey.Contract.keyJsonDecoder();
```

Comment:
Would it make sense to invoke the decoders on concrete values to test that they do their job?

Reply:
Yes, I added `keyToJson` and `keyFromJson` and changed the tests to round-trip.

The code-generation can be reduced a bit by moving some of it into the java bindings, but since that's over in the canton repo, I think I'll need to do those changes separately and wait for the sync to consolidate them.